### PR TITLE
add notification

### PR DESCRIPTION
**Pull request by uriels777**
Hey,
I changed a few files in order to add notifications when an admin add a flair to a user.
didn't add a notification for the automatic flairs, yet.

### CODE REVIEW

File: controller/acp_user_controller.php [hunk 1]
```diff
@@ -258,6 +273,25 @@ protected function change_flair($user_id, $change)
...
+ 				if (isset($this->config['flair_notification_'. $id]))
+				{
+					$this->config->increment('flair_notification_'. $id, 1);
+				}
+				else
+				{
+					$this->config->set('flair_notification_'. $id, 0);
+				}
+				$notification_id = floatval( $this->config['flair_notification_'. $notification_id]. '.' .$id);
+				$insert = [
```
Comment (stevotvr) [inline controller/acp_user_controller.php]: Use array() syntax

File: controller/acp_user_controller.php [hunk 2]
```diff
@@ -267,4 +301,21 @@ protected function change_flair($user_id, $change)
...
 	}
+
+	/**
+	 * send notification to the user or group being worked on and log the action.
+	 *
+	 * @param array $data The data for the notification
+	 */
+	private function send_notification($data)
+	{
+		$this->notification_manager->add_notifications('stevotvr.flair.notification.type.flair', [
```
Comment (stevotvr) [inline controller/acp_user_controller.php]: Use array() syntax

File: event/main_listener.php [hunk 3]
```diff
@@ -90,8 +96,9 @@ class main_listener implements EventSubscriberInterface
 	 * @param \stevotvr\flair\operator\trigger_interface $trigger_operator
 	 * @param \stevotvr\flair\operator\user_interface    $user_operator
 	 * @param string                                     $img_path The path to the custom images
+	 * @param manager $notification_manager	 	 		  $notification_manager Notification manager
```
Comment (stevotvr) [inline event/main_listener.php]: Use fully qualified class names in param and var comments

File: event/main_listener.php [hunk 4]
```diff
@@ -79,6 +80,11 @@ class main_listener implements EventSubscriberInterface
 	 */
 	protected $img_path;
 
+	/**
+	 * @var manager
```
Comment (stevotvr) [inline event/main_listener.php]: Use fully qualified class names in param and var comments

File: controller/acp_user_controller.php [hunk 5]
```diff
@@ -267,4 +301,21 @@ protected function change_flair($user_id, $change)
...
+			'user_ids' => $data['user_ids'],
+			'notification_id' => $data['notification_id'],
+			'username' => $data['username'],
+			'flair_name' => $data['flair_name'],
+		],
+		[
```
Comment (stevotvr) [inline controller/acp_user_controller.php]: Use array() syntax

File: controller/acp_user_controller.php [hunk 6]
```diff
@@ -61,15 +71,19 @@ class acp_user_controller extends acp_base_controller implements acp_user_interf
 	 * @param \stevotvr\flair\operator\category_interface $cat_operator
 	 * @param \stevotvr\flair\operator\flair_interface    $flair_operator
 	 * @param \stevotvr\flair\operator\user_interface     $user_operator
+	 * @param manager $notification_manager	 	 		  $notification_manager Notification manager
```
Comment (stevotvr) [inline controller/acp_user_controller.php]: Use fully qualified class names in param and var comments

File: controller/acp_user_controller.php [hunk 7]
```diff
@@ -52,6 +53,15 @@ class acp_user_controller extends acp_base_controller implements acp_user_interf
 	 */
 	protected $user_operator;
 
+	/**
+	 * @var manager
+	 */
+	private $notification_manager;
+
+
+	/** @var \phpbb\log\log */
```
Comment (stevotvr) [inline controller/acp_user_controller.php]: Check formatting

File: controller/acp_user_controller.php [hunk 8]
```diff
@@ -52,6 +53,15 @@ class acp_user_controller extends acp_base_controller implements acp_user_interf
 	 */
 	protected $user_operator;
 
+	/**
+	 * @var manager
```
Comment (stevotvr) [inline controller/acp_user_controller.php]: Use fully qualified class names in param and var comments

File: event/main_listener.php [hunk 9]
```diff
@@ -166,6 +175,88 @@ public function user_setup(data $event)
...
+		}
+	}
+
+	/**
+	 * Mark all notifications as read
+	 * @param int $post_time
+	 */
+	private function mark_all_read($post_time)
+	{
+		$this->notification_manager->mark_notifications([
```
Comment (stevotvr) [inline event/main_listener.php]: Use array() syntax

File: notification/type/flair.php [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+/**
+* flair notifications class
+* This class handles notifications for flair
+*
+* @package notifications
+*/
+class flair extends \phpbb\notification\type\base
+{
+	/** @var \phpbb\controller\helper */
```
Comment (stevotvr) [inline notification/type/flair.php]: Check formatting

File: notification/type/flair.php [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+	protected $helper;
+
+	/**
+	 * @var config
```
Comment (stevotvr) [inline notification/type/flair.php]: Use fully qualified class names in param and var comments

File: notification/type/flair.php [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+	/**
+	 * @var \phpbb\user_loader
+	 */
+	protected $user_loader;
+	/**
+
+	 * Set the controller helper
+	 *
+	 * @param \phpbb\controller\helper $helper
+	 * @return void
```
Comment (stevotvr) [inline notification/type/flair.php]: You can omit void return types

File: notification/type/flair.php [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+	 *
+	 * @param config $config
+	 *
+	 */
+	public function set_config(config $config)
+	{
+		$this->config = $config;
+	}
+
+	public function set_user_loader(\phpbb\user_loader $user_loader)
```
Comment (stevotvr) [inline notification/type/flair.php]: Missing documentation

File: notification/type/flair.php [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+	}
+
+	/**
+	 * Function for preparing the data for insertion in an SQL query
+	 * (The service handles insertion)
+	 *
+	 * @param array $data The data for the flair
+	 * @param array $pre_create_data Data from pre_create_insert_array()
+	 *
+	 * @return array Array of data ready to be inserted into the database
```
Comment (stevotvr) [inline notification/type/flair.php]: Nothing is returned by this function

File: operator/flair.php [hunk 15]
```diff
@@ -152,4 +152,22 @@ public function count_image_items($image)
 
 		return $count;
 	}
+
+	public function get_flair_name($flair_id)
+	{
+		$flair_name = 'works?';
```
Comment (stevotvr) [inline operator/flair.php]: Leftover debugging?

File: operator/flair_interface.php [hunk 16]
```diff
@@ -77,4 +77,13 @@ public function get_assigned_groups($flair_id);
...
 	public function count_image_items($image);
+
+	/**
+	 * Get the name of this flair item.
+	 *
+	 * @param int $flair_id The database ID of the flair item
+	 *
+	 * @return string The name of this flair item.
+	 */
+	public function get_flair_name($flair_id);
```
Comment (stevotvr) [inline operator/flair_interface.php]: Whitespace found at end of line

File: operator/flair.php [hunk 17]
```diff
@@ -152,4 +152,22 @@ public function count_image_items($image)
...
+		$result = $this->db->sql_query($sql);
+		$row = $this->db->sql_fetchrow($result);
+		if ($row )
+		{
+			$flair_name = $row['flair_name'];
+		}
+		$this->db->sql_freeresult($result);
+
+		return $flair_name;
+	}
```
Comment (stevotvr) [inline operator/flair.php]: Whitespace found at end of line

File: ext.php [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+				$phpbb_notifications->purge_notifications('stevotvr.flair.notification.type.flair');
+				return 'notifications';
+			break;
+			default:
+				// Run parent purge step method
+				return parent::purge_step($old_state);
+			break;
+		}
+	}
+}
```
Comment (stevotvr) [inline ext.php]: Whitespace found at end of line

File: controller/acp_user_controller.php [hunk 19]
```diff
@@ -258,6 +273,25 @@ protected function change_flair($user_id, $change)
 				$this->user_operator->set_flair_count($user_id, $id, 0);
 				return;
 			}
+			else if ($change === 'add')
+			{
+ 				if (isset($this->config['flair_notification_'. $id]))
```
Comment (stevotvr) [inline controller/acp_user_controller.php]: Check indentation

File: notification/type/flair.php [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+ *
+ * @copyright (c) 2018, Steve Guidetti, https://github.com/stevotvr
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ */
+
+namespace stevotvr\flair\notification\type;
+
+use phpbb\config\config;
+use phpbb\notification\type\base;
```
Comment (stevotvr) [inline notification/type/flair.php]: Unused

File: controller/acp_user_controller.php [hunk 21]
```diff
@@ -267,4 +301,21 @@ protected function change_flair($user_id, $change)
...
+	{
+		$this->notification_manager->add_notifications('stevotvr.flair.notification.type.flair', [
+			'user_ids' => $data['user_ids'],
+			'notification_id' => $data['notification_id'],
+			'username' => $data['username'],
+			'flair_name' => $data['flair_name'],
+		],
+		[
+			'user_ids' => $data['user_ids'],
+		]);
```
Comment (stevotvr) [inline controller/acp_user_controller.php]: Missing closing brace

**Review by stevotvr — CHANGES_REQUESTED** [review]
Thanks for your contribution! I added some comments on some issues that need to be addressed.

**Review by stevotvr — APPROVED** [review]

**Comment by koraldon** [comment]
Awesome for the approval :) any idea when you will merge it?
Also, do you think a notification for automatic badges is needed - I assume in any case it is better to have as a separate PR.

**Comment by stevotvr** [comment]
Has this been tested? I'm currently waiting for 1.1.0 to be approved for the extension database before I merge anything.

As for automatic badges, I think it might be useful to be notified about those as well.

**Comment by koraldon** [comment]
We did some testing on this... Waiting to be merged prior to testing on live site...
As it seems it takes a long time for phpbb to approve the extension 1.1., can you merge it before, so we can make progress?

**Comment by stevotvr** [comment]
Alright I merged it into a new branch. Sorry about the wait.

**Comment by stevotvr** [comment]
I noticed in the license header you put the name "Example." Was that intended?

**Comment by uriels777** [comment]
Not a mistake, but I don't really care. You can delete it if you want.
If you decide to add notification for group assignments I think you should disable them when a user enter a group ( notification when you add flair to the group and not when a user enter a group and recive the flairs).
Do you need help with one of the other enhancements?

**Comment by stevotvr** [comment]
So your name is Example?

Anyway, yeah that is probably how the group thing would work since those items are passively assigned via a table join. The notification would be triggered by the admin adding the group.

I think I have the other features covered. I should have it done in a few days.

**Comment by uriels777** [comment]
My name is Uriel if you want to use it. 
The code was an example, so..

**Comment by stevotvr** [comment]
Okay that makes sense.

**Comment by uriels777** [comment]
Realy!?
Sometimes I even confuse myself.

**Pull request merged**